### ISSUE TITLE
meaningless results for road image generation

### ISSUE THREAD

**Issue by WannaSeaU** [lucidrains/stylegan2-pytorch#34]
![image](https://user-images.githubusercontent.com/23076222/77321454-2cd09d00-6d4d-11ea-98cc-1247d2ffbbed.png)
I have trained for 50 epochs on a dataset with 1350 pics, the training data look like the right part of this picture. 
However, the intermediate results (left part of the picture) looks awful. I'm wondering if that stylegan can only generate highly-structured pics?
Thanks in advance.

**Comment by WannaSeaU** [comment]
The command used is `stylegan2_pytorch --data ~/RoadData-Normal --image-size 128 --batch-size 20`.

**Comment by lucidrains** [comment]
@WannaSeaU 1k isn't enough samples, but you should check all of the images that were saved to your results folder. the generator collapses at some point, and you may need to pick the model before it does, not the model at the very end of training

**Comment by WannaSeaU** [comment]
Thanks! The best result I can find is like this. Distortion happens.
![image](https://user-images.githubusercontent.com/23076222/77392567-37834480-6dd6-11ea-9ab5-8cdfa333c9f4.png)
Maybe I should consider enlarging the dataset. Follow is the result of using a DCGAN. It's more like real images but is low-resolution (64*64).
![image](https://user-images.githubusercontent.com/23076222/77392743-98128180-6dd6-11ea-8549-34704bbe6e2c.png)
I'm not familiar with GAN theories, could you please give me some advice? I'm trying to generate road images. Thanks!

**Comment by lucidrains** [comment]
@WannaSeaU 1. get more data! at least 10k. 2. upgrade the framework! there was a bug with the data augmentation and it should be less distorted

**Comment by lucidrains** [comment]
@WannaSeaU wow, it learned the crisscross pattern on the sidewalk lol

**Comment by WannaSeaU** [comment]
Thanks! I'll try the latest version again and seek more data.

**Comment by hackf5** [comment]
I've been training on about 80k images @ 256, batch size 1, gradient-accumulate-every 15. So far I've got through about 80k steps and what I'm seeing is semi-structured noise, there is no human discernible structure starting to form.

The same dataset started to produce decent results on a stylegan 1 after about 12 hours, although I was running at 128. So I guess the dataset is viable, however there is quite a lot of variety.

I'm wondering if this sounds reasonable, like, maybe I'll need to run for 200k+ steps and it still producing junk at 80k is fine, or whether this sounds dodgy and maybe I should start again at 128 and see what it manages, as presumable that will run in around a 1/4 of the time, before going back to 256?

**Comment by hackf5** [comment]
I re-ran with the unmodified code and after around 30k steps I ended up with NAN for the generator loss and the intermediate images were all black. I wasn't getting anything remotely promising out before then though.

**Comment by lucidrains** [comment]
@hackf5 yes, I need to work on saving the training from nans. after playing around with training myself, I realized that a previous commit where I modified the gradient penalty loss seems to have made training worse for images greater than 128x128. I have uploaded a new version and welcome you to try again!

**Comment by lucidrains** [comment]
@WannaSeaU ^^ relevant to you as well!

**Comment by hackf5** [comment]
That sounds good. I've run the dataset about as far as it will go on a StyleGan 1 implementation and the results are OK, but I'd like to see what SG2 can do. However, I think there is a bit too much variation in the data, and I think before doing any more training I'm going to sanitize it a bit to improve the overall quality.

Out of interest how many images would you expect it to need to process before you started seeing something useful?

I can run an entire epoch in about 20 minutes on SG1, but I think it will take a few days to run an epoch on SG2! The model is so much bigger (like 1GB vs 10MB), I can only manage a batch-size of 2 on my GPU.

And what would you suggest for the values of gradient_accumulate_every and learning_rate at a batch size of 2?

Thanks

**Comment by lucidrains** [comment]
@hackf5 I would do a gradient accumulate every of at least `8` and a learning rate of `1e-4`

**Comment by lucidrains** [comment]
@hackf5 You can always try tuning down the model size by decreasing the `--network-capacity` value, but it comes at a cost to the final quality. Also, this isn't the official, optimized repository, it's meant as a concise replication of the architecture. You can always use Nvidia's official tensorflow version if you would like to train faster.

**Comment by hackf5** [comment]
@lucidrains Thanks, I'll give those settings a try. I like PyTorch as it seems more stable on Windows than TF with the new RTX cards. However, I've only tried an unofficial TF2 implementation, so I will give the official TF1 implementation a go in the next couple of days too. TF2 doesn't seem totally stable yet.

**Comment by hackf5** [comment]
I've had a look at the official implementation, but it will take quite a lot of Anaconda environment monkey patching to get it to work on Windows because there is no `nvcc` package available. Half an hour of monkey patching was not enough.

Anyway I've curated my dataset now so it has less variation and I'm re-running on your latest changeset. I'd already got rid of the image mangling code, so I don't think that was the culprit.

I will let you know how I get on. First epoch is probably going to take about 3 days.

**Comment by hackf5** [comment]
So far I've run 70k steps at 256, batch-size 2, grad-accumulate 8, so its seen over a million images, and the results are promising. The images it is generating contain structure. It's got way to go, but so far so good.

Thanks!

**Comment by WannaSeaU** [comment]
I havnt got acceptable results for some reasons, but I'm not working on this currently. Given that @hackf5 has got promising progress, I'm closing this issue now.
